### PR TITLE
Update EIP-663: Move to Draft

### PR DESCRIPTION
**Pull request by charles-cooper**
When opening a pull request to submit a new EIP, please use the suggested template: https://github.com/ethereum/EIPs/blob/master/eip-template.md

We have a GitHub bot that automatically merges some PRs. It will merge yours immediately if certain criteria are met:

 - The PR edits only existing draft PRs.
 - The build passes.
 - Your GitHub username or email address is listed in the 'author' header of all affected PRs, inside <triangular brackets>.
 - If matching on email address, the email address is the one publicly listed on your GitHub profile.

### CODE REVIEW

File: EIPS/eip-663.md
Comment (shemnon) [inline EIPS/eip-663.md:89]: need upper range.  n+m?

Comment (charles-cooper) [reply EIPS/eip-663.md]: added in https://github.com/ethereum/EIPs/pull/7819/commits/e28a856469803c11b0f41f58d1e8bd2dbb1fe71a

**Comment by charles-cooper** [comment]
if stack depth addressability is a concern, `EXCHANGE2` can be added (which can address at least up to 512 stack depth)

**Comment by ekpyron** [comment]
This is not a reasonable modification to EIP-663. Note that with this PR, the EIP's content does not even match the EIP title anymore.
You're trying to solve quite a different problem with `EXCHANGE`, so this seems more like hijacking the EIP to fully repurpose it, rather than an amendment :-).

I do appreciate that there are reasons why one may want an ``EXCHANGE`` opcode - stack scheduling does require more creative algorithms without it - but I'd argue that it should be proposed as a separate EIP. Other than both having something to do with stack, there is little relation between EIP-663 and the proposed `EXCHANGE`. I'd even be in support of such a new EIP, but as an (for us optional) addition to, not as replacement of, EIP-663 (and contingently on the acceptance of such a new EIP, EIP-663 could then be amended by an *additional* `EXCHANGE2` - or conversely the new EIP should add that, if EIP-663 is accepted).

**Comment by charles-cooper** [comment]
> This is not a reasonable modification to EIP-663. Note that with this PR, the EIP's content does not even match the EIP title anymore. You're trying to solve quite a different problem with `EXCHANGE`, so this seems more like hijacking the EIP to fully repurpose it, rather than an amendment :-).
> 
> I do appreciate that there are reasons why one may want an `EXCHANGE` opcode - stack scheduling does require more creative algorithms without it - but I'd argue that it should be proposed as a separate EIP. Other than both having something to do with stack, there is little relation between EIP-663 and the proposed `EXCHANGE`. I'd even be in support of such a new EIP, but as an (for us optional) addition to, not as replacement of, EIP-663 (and contingently on the acceptance of such a new EIP, EIP-663 could then be amended by an _additional_ `EXCHANGE2` - or conversely the new EIP should add that, if EIP-663 is accepted).

sorry, i don't think "hijacking" is a fair characterization here. i have been bringing this up for nearly a year now on the discussion thread (https://ethereum-magicians.org/t/eip-663-unlimited-swap-and-dup-instructions/3346/40), and regularly attending EOF meetings, and the suggestion was made that i make a PR directly against EIP-663.

additionally, i didn't add different variants (and there are several variants which i think are useful) of SWAP/EXCHANGE since i have been informed that opcode slot space is at a premium. so i tried to make a minimal change. but (and maybe this is just context from regularly attending and participating in the EOF calls, but did not come across expressed clearly in the PR) `EXCHANGE` and its `EXCHANGE2` variant (with 2-byte immediate) fully subsume the functionality of the original proposed `SWAPN` opcode, with an opcode slot vs bytecode size tradeoff.

**Comment by ekpyron** [comment]
> sorry, i don't think "hijacking" is a fair characterization here. i have been bringing this up for nearly a year now on the discussion thread (https://ethereum-magicians.org/t/eip-663-unlimited-swap-and-dup-instructions/3346/40), and regularly attending EOF meetings, and the suggestion was made that i make a PR directly against EIP-663.
> 
> additionally, i didn't add different variants (and there are several variants which i think are useful) of SWAP/EXCHANGE since i have been informed that opcode slot space is at a premium. so i tried to make a minimal change. but (and maybe this is just context from regularly attending and participating in the EOF calls, but did not come across expressed clearly in the PR) `EXCHANGE` and its `EXCHANGE2` variant (with 2-byte immediate) fully subsume the functionality of the original proposed `SWAPN` opcode, with an opcode slot vs bytecode size tradeoff.

I apologize for the phrasing, I didn't mean to insinuate misintent - and I know that Vyper has been in favour of an exchange opcode in EOF for a long time. But I want it to be as clear as possible that this PR as it is stated here (which markedly does *not* include ``EXCHANGE2``) is not an option. It clearly changes the EIP in such a way that it no longer solves the issue this EIP is meant to address, i.e. limited reach on stack.

If this is to be proposed as amendment of EIP-663 and not as a separate proposal, it has to keep addressing this problem, which the current state of the PR doesn't. A variant of the PR that introduces `EXCHANGE2` instead or additionally would be a different story.

File: EIPS/eip-663.md
Comment (ekpyron) [inline EIPS/eip-663.md:79]: This argument here also technically doesn't hold up to scrutiny.
Proper EVM stack scheduling can avoid situations in which you need to exchange two slots neither of which are at the stack top. I.e. if you find yourself in the situation of having to swap the 3rd and 2nd position on stack, then you allocated an inefficient position for the values on stack when you produced them in the first place. I've given a talk at Devconnect Amsterdam about the topic, if you're interested.
Conversely, all values in the EVM are *produced* at the stack top and it thereby cannot be avoided to swap the stack top with a deeper position to create optimal stack layouts in the first place.

So, technically, `SWAPN` is actually the more efficient choice, since the more common and unavoidable operation on stack top requires only two bytes instead of three even for deep access, while conversely patterns like `SWAP2 SWAP3 SWAP2` can generally be avoided by proper stack scheduling at the compiler level.

That being said, having written stack scheduling like that myself, I know that it is not the most trivial thing to do - that's why I've not immediately tried to shut down the idea of `EXCHANGE`: it does allow for simpler more naive stack scheduling albeit at the cost of an increase in code size cost globally compared to optimal scheduling.

But either way, `DUPN` and `EXCHANGE` are not a replacement of `SWAPN`, so it would have to be `EXCHANGE2` as well, no matter what.

Comment (charles-cooper) [reply EIPS/eip-663.md:79]: > generally be avoided by proper stack scheduling at the compiler level.

the situation can definitely be improved with better algorithms, but optimal stack scheduling, especially across basic blocks and subroutines, is hard. (although i don't have a proof, i suspect it is NP-complete, in the same way that register allocation is). and, even if such a theoretically optimal stack scheduler were available, it is sometimes simply not possible in the general case to schedule things to be in the right order ahead of time, for instance, when side effects are interleaved with stack item production.

> But either way, DUPN and EXCHANGE are not a replacement of SWAPN, so it would have to be EXCHANGE2 as well, no matter what.

having things too deep in the stack is also solvable by better memory spilling algorithms, so i don't think phrasing this choice in absolute terms is very constructive here. but sure, if `EXCHANGE` + `EXCHANGE2` are not considered as taking up too many slots, why not.

**Comment by charles-cooper** [comment]
> Note that with this PR, the EIP's content does not even match the EIP title anymore.

the original EIP itself did not match the EIP title! the title implies "unlimited" reach, but the opcodes only provide 256- slots of addressability. (and even the three-byte variant, of course, only allows addressing up to the 1024 limit imposed by the EVM). so clearly there is a limit, but it is deemed practically useful. the proposed `EXCHANGE` opcode also increases addressability, the same as the original EIP, it just increases the reach somewhat less (in exchange for the other useful semantics as already described).

**Comment by ekpyron** [comment]
The easiest thing to do and to avoid having to senselessly split hairs here would be to just make the introduction of ``EXCHANGE`` a separate EIP, since it's obviously a radically different change compared to the unmodified EIP-663 (you cannot reasonably deny that).
As I said, I'd grant that that EIP would have (independent) merit.

**Comment by charles-cooper** [comment]
modified the PR so it adds `EXCHANGE` in addition to(!) `SWAPN`.

File: EIPS/eip-663.md
Comment (pdobacz) [inline EIPS/eip-663.md:55]: Shouldn't this be `n+1`th? `SWAPN 0x00` is then equivalent to `SWAP1` and swaps 2nd (1-based) item with top 1st item. 

Note that `DUPN 0x00` is equivalent to `DUP1`.

Similarly `EXCHANGE 0x00` is equivalent to `SWAP1`

File: EIPS/eip-663.md
Comment (gumb0) [inline EIPS/eip-663.md:56]: This limits the swapping of `nth` item to only items `[n+1, n+16]`, so it can go 32 items deep but only with max distance 16 between items. E.g. you can't swap 1st with 32d.

Is this acceptable limitation for you?

File: EIPS/eip-663.md
Comment (charles-cooper) [reply EIPS/eip-663.md]: thanks, resolved. i must have messed those up when merging different commits i had

File: EIPS/eip-663.md
Comment (charles-cooper) [reply EIPS/eip-663.md]: yes, we might want to change the encoding still.

File: EIPS/eip-663.md
Comment (pdobacz) [inline EIPS/eip-663.md:92]: I think my other comment got lost: `SWAPN` cases got lost in merges, this entire section needs a revisit at the end. Also specifically it's `n + m + 1` now for EXCHANGE

File: EIPS/eip-663.md
Comment (pdobacz) [inline EIPS/eip-663.md:51]: `n + m + 1`

**Review by pdobacz — COMMENTED** [review]
A few more off-by-one nitpicks, since we increased the reach of EXCHANGE. Also the stack depths mentioned in the abstract aren't 100% accurate, but maybe that's intended to not overwhelm with detail

File: EIPS/eip-663.md
Comment (gumb0) [inline EIPS/eip-663.md:36]: Conflicts with current propsoal for `DATALOAD`

Comment (pdobacz) [reply EIPS/eip-663.md:36]: Have we chosen an alternative yet? Need to fix it in the megaspec (https://github.com/ipsilon/eof/issues/43)

Comment (pdobacz) [reply EIPS/eip-663.md:36]: proposing moving a bunch of new opcodes to the D section here: https://github.com/ipsilon/eof/pull/46

Comment (pdobacz) [reply EIPS/eip-663.md:36]: DATALOAD is being moved to 0xd0

File: EIPS/eip-663.md
Comment (pdobacz) [inline EIPS/eip-663.md:16]: `32, 32 and 256` sounds off, shouldn't be `257, 256 and 33`? - however I'm wondering if this isn't too detailed. Maybe can consider making it vague "manipulating stack items at much greater depth"

File: EIPS/eip-663.md
Comment (pdobacz) [inline EIPS/eip-663.md:7]: 3540 got marked as stagnant (will be fixed soon). Can we in the meantime put this back to `Draft`? Otherwise we can't merge the PR :(

File: EIPS/eip-663.md
Comment (pdobacz) [inline EIPS/eip-663.md:92]:
```suggestion
    - `SWAPN imm` to fail validation if `stack_height < n + 1`.
    - `EXCHANGE imm` to fail validation if `stack_height < n + m + 1`.
```



bring back the cases for SWAPN which got lost

File: EIPS/eip-663.md
Comment (pdobacz) [inline EIPS/eip-663.md:94]:
```suggestion
    - `DUPN imm`, `SWAPN imm`, and `EXCHANGE imm` to fail at run-time if gas available is less than 3.
```

File: EIPS/eip-663.md
Comment (pdobacz) [inline EIPS/eip-663.md:95]:
```suggestion
    - otherwise `DUPN imm` should push the `stack[n]` item to the stack, and `SWAPN imm` should swap `stack[n + 1]` with `stack[stack.top()]` and `EXCHANGE imm` should swap `stack[n + 1]` with `stack[n + m + 1]`.
```

**Review by pdobacz — COMMENTED** [review]
Please double check all the indices suggested

**Review by pdobacz — APPROVED** [review]

File: EIPS/eip-663.md
Comment (gumb0) [inline EIPS/eip-663.md:7]: We hope to get EIP-3540 out of Stagnant soon https://github.com/ethereum/EIPs/pull/8217
then this can be left in Review

Comment (pdobacz) [reply EIPS/eip-663.md:7]: @charles-cooper :point_up: this can be undone now, 3540 is out of Stagnant.

File: EIPS/eip-663.md
Comment (SamWilsn) [inline EIPS/eip-663.md:4]: Same kind of comment for the description.

**Review by SamWilsn — APPROVED** [review]
This'll still need approval from @axic to add the new authors.

File: EIPS/eip-663.md
Comment (pdobacz) [reply EIPS/eip-663.md:3]: :+1: Andrei's proposal. But also - would just bringing the mnemonics back into the title / description work ("`SWAPN`, `DUPN`, `EXCHANGE`")?

File: EIPS/eip-663.md
Comment (gumb0) [inline EIPS/eip-663.md:96]: These are off by one I think, because indexing of `n` is 1-based.
```suggestion
    - otherwise `DUPN imm` should push the `stack[n - 1]` item to the stack, and `SWAPN imm` should swap `stack[n]` with `stack[stack.top()]` and `EXCHANGE imm` should swap `stack[n]` with `stack[n + m]`.
```
Maybe better to just omit repeating the spec another time here.

File: EIPS/eip-663.md
Comment (gumb0) [inline EIPS/eip-663.md:100]: maybe
```suggestion
The authors are not aware of any additional risks introduced here. The EVM stack is fixed at 1024 items and most implementations keep that in memory at all times. This change will increase the number of easy-to-access stack items.
```

File: EIPS/eip-663.md
Comment (charles-cooper) [reply EIPS/eip-663.md:96]: so just remove line 96?

Comment (charles-cooper) [reply EIPS/eip-663.md:96]: yea, why do the test cases simply repeat the spec -- we should have actual test cases

**Review by axic — APPROVED** [review]

**Pull request merged**
